src/pint/models/stand_alone_psr_binaries/BT_piecewise.py

```diff
@@ -0,0 +1,441 @@
+            self.group_index_array = self.toa_belongs_in_group(t)
+        if len(self.group_index_array) != len(t):
+            self.group_index_array = self.toa_belongs_in_group(t)
+        for i in self.group_index_array:
+            if i != -1:
+                for j in range(len(self.piecewise_parameter_information)):
+                    if (
+                        self.piecewise_parameter_information[j][0] == i
+                    ):  # searches the 5 x n array to find the index matching the toa_index
+                        T0X_per_toa.append(
+                            self.piecewise_parameter_information[j][1].value
+                        )
+                        A1X_per_toa.append(
+                            self.piecewise_parameter_information[j][2].value
+                        )
+            else:  # if a toa lies between 2 groups, use default T0/A1 values (i.e. toa lies after previous upper bound but before next lower bound)
+                T0X_per_toa.append(self.T0.value)
+                A1X_per_toa.append(self.A1.value)
+        T0X_per_toa = T0X_per_toa * u.d
+        A1X_per_toa = A1X_per_toa * ls
+        return [T0X_per_toa, A1X_per_toa]
+
+    def group_index(self):
+        index = []
+        for i in range(len(self._t)):
+            index1 = self.lower_index[i]
+            index2 = self.upper_index[i]
+            if index1 == index2:
+                index.append(index1)
+            else:
+                index.append(-1)
+        self.group_index_array = np.array(index)
+        return self.group_index_array
+
+    def toa_belongs_in_group(self, t):
+        group_no = []
+        lower_edge, upper_edge = self.get_group_boundaries()
+        for i in t:
+            # print(lower_edge)
+            lower_bound = np.searchsorted(lower_edge, i) - 1
+            upper_bound = np.searchsorted(upper_edge, i)
+            if lower_bound == upper_bound:
+                index_no = lower_bound
+            else:
+                index_no = -1
+
+            if index_no != -1:
+                group_no.append(self.piecewise_parameter_information[index_no][0])
+            else:
+                group_no.append(index_no)
+        # print(group_no)
+        return group_no
+
+    def get_group_boundaries(self):
+        lower_group_edge = []
+        upper_group_edge = []
+        # print("from get_group_boundaries")
+        if hasattr(self, "piecewise_parameter_information"):
+            for i in range(0, len(self.piecewise_parameter_information)):
+                lower_group_edge.append(
+                    self.piecewise_parameter_information[i][3].value
+                )
+                upper_group_edge.append(
+                    self.piecewise_parameter_information[i][4].value
+                )
+            return [lower_group_edge, upper_group_edge]
+
+    def a1(self):
+        self.A1_val = self.A1X_arr * ls
+        if hasattr(self, "A1X_per_toa") is True:
```

Comment:
What decides whether `self` has `A1X_per_toa`? When would it happen or not? Can you make sure it's always present?